core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,772 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020, Arm Limited. All rights reserved.
+ * Copyright (c) 2019, Linaro Limited
+ */
+
+#include <kernel/abort.h>
+#include <kernel/secure_partition.h>
+#include <kernel/user_mode_ctx.h>
+#include <mm/fobj.h>
+#include <mm/mobj.h>
+#include <mm/tee_mmu.h>
+#include <pta_stmm.h>
+#include <tee/tee_svc.h>
+#include <tee/tee_pobj.h>
+#include <zlib.h>
+
+#include <ffa.h>
+#include "thread_private.h"
+#include <tee/tee_svc_storage.h>
+#include <crypto/crypto.h>
+#include <tee_api_defines_extensions.h>
+
+static const TEE_UUID stmm_uuid = PTA_STMM_UUID;
+
+/*
+ * Once a complete FFA spec is added, these will become discoverable.
+ * Until then these are considered part of the internal ABI between
+ * OP-TEE and StMM
+ */
+static const uint16_t stmm_id = 1U;
+static const uint16_t stmm_pta_id = 2U;
+static const uint16_t mem_mgr_id = 3U;
+static const uint16_t ffa_storage_id = 4U;
+
+static const unsigned int stmm_stack_size = 4 * SMALL_PAGE_SIZE;
+static const unsigned int stmm_heap_size = 398 * SMALL_PAGE_SIZE;
+static const unsigned int stmm_sec_buf_size = SMALL_PAGE_SIZE;
+static const unsigned int stmm_ns_comm_buf_size = SMALL_PAGE_SIZE;
+
+extern uint8_t stmm_image[];
+extern const unsigned int stmm_image_size;
+extern const unsigned int stmm_image_uncompressed_size;
+
+static struct sec_part_ctx *sec_part_alloc_ctx(const TEE_UUID *uuid)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sec_part_ctx *spc = NULL;
+
+	spc = calloc(1, sizeof(*spc));
+	if (!spc)
+		return NULL;
+
+	spc->uctx.ctx.ops = &secure_partition_ops;
+	spc->uctx.ctx.uuid = *uuid;
+	spc->uctx.ctx.flags = TA_FLAG_SINGLE_INSTANCE |
+			      TA_FLAG_INSTANCE_KEEP_ALIVE;
+
+	res = vm_info_init(&spc->uctx);
+	if (res)
+		goto err;
+
+	spc->uctx.ctx.ref_count = 1;
+	condvar_init(&spc->uctx.ctx.busy_cv);
+
+	return spc;
+err:
+	free(spc);
+	return NULL;
+}
+
+static void clear_vfp_state(struct sec_part_ctx *spc __maybe_unused)
+{
+#ifdef CFG_WITH_VFP
+	thread_user_clear_vfp(&spc->uctx.vfp);
+#endif
+}
+
+static TEE_Result sec_part_enter_user_mode(struct sec_part_ctx *spc)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t exceptions = 0;
+	uint32_t panic_code = 0;
+	uint32_t panicked = 0;
+	uint64_t cntkctl = 0;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cntkctl = read_cntkctl();
+	write_cntkctl(cntkctl | (CNTKCTL_PL0PCTEN | CNTKCTL_PL0VCTEN));
+	__thread_enter_user_mode(&spc->regs, &panicked, &panic_code);
+	write_cntkctl(cntkctl);
```

Comment:
Do we need to restore any counters too, in case the SP has done something funny?

Reply:
I'll dig into StMM a bit further for this. 
There's something I'll need to take a closer look as well. 
When I tried that on actual hardware, we got an error since we were unable to access those. 
Trying now on QEMU everything seems to work fine. So I'll dopuble check and see if some StMM code got removed in the meanwhile and this isn't needed anymore

Reply:
Ok first of all CNTKCTL_PL0PCTEN is only needed. StMM does not perform any writes on it, so it's fine for now.
Taking into account the bigger picture though of running multiple SPs. 
The only register that can change on that is CNTFRQ, should we just check if that changed and try to restore it or warn?

Reply:
A biot more digging. CNTPCT is read only so we wont have to bother, CNTFRQ is only writable from the highest EL level, so an SP will not be able to change that

Reply:
And this is what you get when you try writing CNTFRQ from S-EL0
OP-TEE: revision 3.9 (7de6385f)
E/TC:? 0 
E/TC:? 0 User TA undef-abort at address 0x4040bc00
E/TC:? 0  esr 0x02000000  ttbr0 0x200009e93e000   ttbr1 0x00000000   cidr 0x0
E/TC:? 0  cpu #0          cpsr 0x60000000
E/TC:? 0  x0  000000001dcd6500 x1  0000000040411130
E/TC:? 0  x2  0000000040411c40 x3  00000000404111b8
E/TC:? 0  x4  0000000040411c40 x5  00000000404111b0
E/TC:? 0  x6  00000000404111a8 x7  0000000040415c48
E/TC:? 0  x8  4c01bd31d326d041 x9  492eb68926eeb3de
E/TC:? 0  x10 0000000040411190 x11 0000000000000001
E/TC:? 0  x12 000000004000b2b0 x13 000000004000b2b0
E/TC:? 0  x14 000000004040a000 x15 000000004040c010
E/TC:? 0  x16 0000000000000000 x17 0000000000000000
E/TC:? 0  x18 0000000000000000 x19 0000000000000000
E/TC:? 0  x20 0000000040411510 x21 000000004040afb0
E/TC:? 0  x22 0000000040415e68 x23 0000000040415ef8
E/TC:? 0  x24 0000000040415e68 x25 0000000040415ef8
E/TC:? 0  x26 0000000000000000 x27 000000004040a000
E/TC:? 0  x28 000000004040ae58 x29 0000000040415cc0
E/TC:? 0  x30 000000004040bb9c elr 000000004040bc00
E/TC:? 0  sp_el0 0000000040415cc0
E/TC:? 0  region  0: va 0x0000000040000000 pa 0x000000009e802000 size 0x002000 flags ---R-X
E/TC:? 0  region  1: va 0x0000000040002000 pa 0x000000009e930000 size 0x001000 flags ---RW-
E/TC:? 0  region  2: va 0x0000000040004000 pa 0x000000009ea00000 size 0x001000 flags r-xR--
E/TC:? 0  region  3: va 0x0000000040005000 pa 0x000000009ea01000 size 0x001000 flags rw----
E/TC:? 0  region  4: va 0x0000000040006000 pa 0x000000009ea02000 size 0x005000 flags r-x---
E/TC:? 0  region  5: va 0x000000004000b000 pa 0x000000009ea07000 size 0x001000 flags rw----
E/TC:? 0  region  6: va 0x000000004000c000 pa 0x000000009ea08000 size 0x001000 flags r-x---
E/TC:? 0  region  7: va 0x000000004000d000 pa 0x000000009ea09000 size 0x277000 flags r-xR--
E/TC:? 0  region  8: va 0x0000000040284000 pa 0x000000009ec80000 size 0x187000 flags rw-RW-
E/TC:? 0  region  9: va 0x000000004040b000 pa 0x000000009ee07000 size 0x001000 flags r-x---
E/TC:? 0  region 10: va 0x000000004040c000 pa 0x000000009ee08000 size 0x00b000 flags rw-RW-
E/TC:? 0  region 11: va 0x0000000040417000 pa 0x000000009ee13000 size 0x001000 flags rw-RW-
EFI: cannot initialize platform language settings
Error: Cannot initialize UEFI sub-system, r = 2